File: resources/patterns/design/repository-pattern/user_repository.py
"""Repository Pattern — Python implementation."""
from abc import ABC, abstractmethod
from typing import Optional


class User:
    def __init__(self, id: int, name: str, role: str = "member"):
        self.id = id
        self.name = name
        self.role = role

    def __eq__(self, other):
        return isinstance(other, User) and self.id == other.id

    def __repr__(self):
        return f"User(id={self.id}, name={self.name!r}, role={self.role!r})"


class UserRepository(ABC):
    """Abstract repository interface — the contract."""

    @abstractmethod
    def get_by_id(self, id: int) -> Optional[User]:
        ...

    @abstractmethod
    def find_all(self) -> list[User]:
        ...

    @abstractmethod
    def save(self, user: User) -> None:
        ...

    @abstractmethod
    def delete(self, id: int) -> bool:
        ...


class InMemoryUserRepository(UserRepository):
    """In-memory implementation for fast, deterministic unit tests."""

    def __init__(self):
        self._users: dict[int, User] = {}

    def get_by_id(self, id: int) -> Optional[User]:
        return self._users.get(id)

    def find_all(self) -> list[User]:
        return list(self._users.values())

    def save(self, user: User) -> None:
        self._users[user.id] = user

    def delete(self, id: int) -> bool:
        return self._users.pop(id, None) is not None


class UserService:
    """Domain service — depends on the interface, not the concrete class."""

    def __init__(self, repo: UserRepository):
        self._repo = repo

    def promote_user(self, id: int) -> User:
        user = self._repo.get_by_id(id)
        if user is None:
            raise ValueError("User not found")
        user.role = "admin"
        self._repo.save(user)
        return user

File: resources/patterns/design/repository-pattern/user_repository.js
// Repository Pattern — JavaScript implementation.

class User {
  constructor(id, name, role = "member") {
    this.id = id;
    this.name = name;
    this.role = role;
  }
}

class UserRepository {
  /** @abstract */
  getById(id) {
    throw new Error("Not implemented");
  }

  /** @abstract */
  findAll() {
    throw new Error("Not implemented");
  }

  /** @abstract */
  save(user) {
    throw new Error("Not implemented");
  }

  /** @abstract */
  delete(id) {
    throw new Error("Not implemented");
  }
}

class InMemoryUserRepository extends UserRepository {
  constructor() {
    super();
    this.users = new Map();
  }

  getById(id) {
    return this.users.get(id) ?? null;
  }

  findAll() {
    return Array.from(this.users.values());
  }

  save(user) {
    this.users.set(user.id, user);
  }

  delete(id) {
    return this.users.delete(id);
  }
}

class UserService {
  /** @param {UserRepository} repo */
  constructor(repo) {
    this.repo = repo;
  }

  promoteUser(id) {
    const user = this.repo.getById(id);
    if (!user) throw new Error("User not found");
    user.role = "admin";
    this.repo.save(user);
    return user;
  }
}

module.exports = { User, UserRepository, InMemoryUserRepository, UserService };
